<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>Hierarchical Models for RNA-seq</title>
  <meta name="description" content="">
  <meta name="author" content="Michael Love">
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />
  <link rel="stylesheet" href="libraries/frameworks/revealjs/css/reveal.min.css">
  <link rel="stylesheet" href="libraries/frameworks/revealjs/css/theme/serif.css" id="theme">
  <link rel="stylesheet" href="libraries/highlighters/highlight.js/css/tomorrow.css" id="theme">
  <!--[if lt IE 9]>
  <script src="lib/js/html5shiv.js"></script>
  <![endif]-->  
</head>
<body>
  <div class="reveal">
    <div class="slides">
      <section class='' data-state='' id='slide-1'>
  <h3>Hierarchical Models for RNA-seq</h3>
  <p><br></p>

<p>Michael Love</p>

<p>Dept of Biostatistics</p>

<p>Dept of Genetics</p>

  
</section>
<section class='' data-state='' id='slide-2'>
  <h3>DNA =&gt; RNA</h3>
  <p><img src="img/tx_elong.png" alt=""></p>

<p><img src="img/tx_em.jpg" alt=""></p>

  
</section>
<section class='' data-state='' id='slide-3'>
  <h3>Why measure RNA as phenotype</h3>
  <p><img src="img/people.jpg" alt=""></p>

  
</section>
<section class='' data-state='' id='slide-4'>
  <h3>Why measure RNA: tissue diversity</h3>
  <p><img src="img/roadmap.jpg" alt=""></p>

<p><small><a href="http://www.roadmapepigenomics.org">Roadmap Epigenomics</a></small></p>

  
</section>
<section class='' data-state='' id='slide-5'>
  <h3>Why measure RNA: tissue diversity</h3>
  <p><img src="img/gtex.png" alt=""></p>

<p><small><a href="http://www.gtexportal.org/home/">GTEx</a></small></p>

  
</section>
<section class='' data-state='' id='slide-6'>
  <h3>Why measure RNA: within tissue over time</h3>
  <p><img src="img/circ.png" alt=""></p>

<p><small><a href="http://www.pnas.org/content/111/45/16219.full">Zhang, et al. Circadian gene expression atlas (2014)</a></small></p>

  
</section>
<section class='' data-state='' id='slide-7'>
  <h3>Why measure RNA:  disease sub-types</h3>
  <p><img src="img/perou.png" alt=""></p>

<p><small><a href="http://www.nature.com/nature/journal/v406/n6797/full/406747a0.html">Perou, et al. Molecular portraits of human breast tumours (2000)</a></small></p>

  
</section>
<section class='' data-state='' id='slide-8'>
  <h3>Step back: pre-sequencing</h3>
  <p>Before sequencing was microarray</p>

<p><img src="img/ma2.png" alt=""> </p>

  
</section>
<section class='' data-state='' id='slide-9'>
  <h3>Step back: pre-sequencing</h3>
  <p>Signal was captured light (positive, &quot;continuous&quot;)</p>

<p><img src="img/ma.jpg" alt=""></p>

  
</section>
<section class='' data-state='' id='slide-10'>
  <h3>Formalize</h3>
  <ul>
<li>Formalize as a simple statistical model</li>
<li>Observed data \(X_{ij}\) as a R.V. following a distribution</li>
<li>Distributional parameters of interest </li>
</ul>

  
</section>
<section class='' data-state='' id='slide-11'>
  <h3>Motivating problem</h3>
  <ul>
<li>Gene expression for <em>i</em>=1,...,N genes and <em>j</em>=1,...,M samples</li>
<li>log of gene expression values are in a tall matrix X</li>
<li>log here is convenient because gene expression is non-negative and has a
long tail</li>
<li>2 equal sized groups of samples A and B</li>
</ul>

<p><br></p>

<p>\[
\begin{aligned}
X_{ij} &\sim N(\mu_{ij}, \sigma_i) \\
\mu_{ij} &= \mu_{i0}, \quad j \in A \\
\mu_{ij} &= \mu_{i0} + \delta_i, \quad j \in B
\end{aligned}
\]</p>

<p><br></p>

<p>\(\delta_i \ne 0\) implies DE (differential expression)</p>

  
</section>
<section class='' data-state='' id='slide-12'>
  <h3>Note \(\sigma_i\)</h3>
  <p>This is <em>critical</em>: different genes <em>i</em> have different amount of variability.</p>

<p><br></p>

<p>\[ 
\begin{aligned}
X_{ij} &\sim N(\mu_{ij}, \sigma_i) \\
\mu_{ij} &= \mu_{i0}, \quad j \in A \\
\mu_{ij} &= \mu_{i0} + \delta_i, \quad j \in B
\end{aligned}
\]</p>

  
</section>
<section class='' data-state='' id='slide-13'>
  <h3>Goal of differential expression</h3>
  <ul>
<li>Find a set of genes for which \(\delta_i \ne 0\)</li>
<li>We want to target a false discovery rate (FDR) cutoff</li>
<li>For genes in our set G at FDR threshold z</li>
</ul>

<p><br></p>

<p>\[
E \left( \sum\nolimits_{i \in G} 1_{ \{\delta_i = 0\} } \right) \le
\left\vert G \right\vert z
\]</p>

  
</section>
<section class='' data-state='' id='slide-14'>
  <h3>Is this realistic?</h3>
  <ul>
<li>Can we accomplish this if all \(\delta_i \ne 0\)

<ul>
<li>no, because methods often rely on global scaling normalization</li>
</ul></li>
<li>Are any \(\delta_i = 0\)? 

<ul>
<li>maybe not, but many are very small for controlled experiment</li>
</ul></li>
</ul>

  
</section>
<section class='' data-state='' id='slide-15'>
  <h3>Is this realistic?</h3>
  <ul>
<li>What about \(\sigma_i\) for both groups?

<ul>
<li>often this is enough, larger variance dominates</li>
<li>not for single cell experiments</li>
</ul></li>
<li>More complex parametric models: <a href="bioconductor.org/packages/baySeq">baySeq</a></li>
<li>Non-parametric: <a href="http://statweb.stanford.edu/%7Etibs/SAM/">SAM / SAMseq</a></li>
</ul>

  
</section>
<section class='' data-state='' id='slide-16'>
  <h3>Back to the model</h3>
  <p>\[
\begin{aligned}
X_{ij} &\sim N(\mu_{ij}, \sigma_i) \\
\mu_{ij} &= \mu_{i0}, \quad j \in A \\
\mu_{ij} &= \mu_{i0} + \delta_i, \quad j \in B
\end{aligned}
\]</p>

<p><br></p>

<ul>
<li>N = 5000, M = 6</li>
<li>\(\delta_i = 0\) for 90%</li>
<li>\(\delta_i = \pm2\) for 10%</li>
<li>\(\sigma_i \sim \Gamma(10,10)\) </li>
</ul>

  
</section>
<section class='' data-state='' id='slide-17'>
  <h3>Distribution of \(\sigma_i\)</h3>
  <p><img src="assets/fig/sigmadist-1.png" alt="plot of chunk sigmadist"></p>

  
</section>
<section class='' data-state='' id='slide-18'>
  <h3>Try simple row t-tests</h3>
  <p><img src="assets/fig/boxt-1.png" alt="plot of chunk boxt"></p>

  
</section>
<section class='' data-state='' id='slide-19'>
  <h3>Just looking at ranks</h3>
  <p><img src="assets/fig/roc-1.png" alt="plot of chunk roc"></p>

  
</section>
<section class='' data-state='' id='slide-20'>
  <h3>Characterize the false positives</h3>
  <p>\[
\textrm{med}(t) \equiv \textrm{median}( \left| t_i \right| )
\, \textrm{for} \, i : \delta_i \ne 0
\]</p>

<p><img src="assets/fig/mediant-1.png" alt="plot of chunk mediant"></p>

  
</section>
<section class='' data-state='' id='slide-21'>
  <h3>Estimates of \(\sigma_i\)</h3>
  <p>\[
\textrm{med}(t) \equiv \textrm{median}( \left| t_i \right| )
\, \textrm{for} \, i : \delta_i \ne 0
\]</p>

<p><img src="assets/fig/fp-1.png" alt="plot of chunk fp"></p>

  
</section>
<section class='' data-state='' id='slide-22'>
  <h3>New estimator for \(\sigma_i\)</h3>
  <p>\[
\begin{aligned}
\bar{\sigma} &= \frac{1}{N} \sum\nolimits_{i=1}^N \hat{\sigma}_i \\
\tilde{\sigma}_i^B &= B \bar{\sigma} + (1-B) \hat{\sigma}_i
\end{aligned}
\]</p>

<p><img src="assets/fig/tildesigma-1.png" alt="plot of chunk tildesigma"></p>

  
</section>
<section class='' data-state='' id='slide-23'>
  <h3>New estimator performance by rank</h3>
  <p><img src="assets/fig/roc2-1.png" alt="plot of chunk roc2"></p>

  
</section>
<section class='' data-state='' id='slide-24'>
  <h3>New estimators by |t|</h3>
  <p><img src="assets/fig/newestmediant-1.png" alt="plot of chunk newestmediant"></p>

  
</section>
<section class='' data-state='' id='slide-25'>
  <h3>Summary</h3>
  <ul>
<li>Top false positives were coming from genes with too low \(\hat{\sigma}_i\)</li>
<li>Replace \(\hat{\sigma}_i\) with an estimate which is closer to
\(\bar{\sigma}\)</li>
<li>Depending on &quot;close&quot;, new estimator dominates at all thresholds</li>
</ul>

  
</section>
<section class='' data-state='' id='slide-26'>
  <h3>How is this hierarchical?</h3>
  <p>Not your standard diagram, need to formalize</p>

<p><img src="img/plate1.png" alt=""></p>

  
</section>
<section class='' data-state='' id='slide-27'>
  <h3>limma</h3>
  <ul>
<li><a href="http://www.statsci.org/smyth/pubs/ebayes.pdf">Smyth, G. K. (2004)</a> 
Linear models and empirical Bayes methods for
assessing differential expression in microarray experiments </li>
<li>Developed the hierarchical model introduced by 
Lonnstedt and Speed (2002) for single sample into method for any
experiment represented as linear model</li>
</ul>

<p><br></p>

<p>\[
\frac{1}{\sigma^2_i} \sim \frac{1}{d_0 \sigma_0^2} \chi_{d0}^2 
\]</p>

<p><br></p>

  
</section>
<section class='' data-state='' id='slide-28'>
  <h3>Why inverse \(\chi^2\)?</h3>
  <ul>
<li>Conjugacy provides closed form solution</li>
<li>Posterior mean for \(1/\sigma_i^2\) given \(\hat{\sigma}_i^2\) is
\(1/\tilde{\sigma}_i^2\) with</li>
</ul>

<p><br></p>

<p>\[
\tilde{\sigma}_i^2 = \frac{d_0 \hat{\sigma}_0^2 + d_i \hat{\sigma}_i^2}{d_0 + d_i} 
\]</p>

<p>And \(d_i\) as the standard residual degrees of freedom</p>

  
</section>
<section class='' data-state='' id='slide-29'>
  <h3>Note that \(d_0\) controls B</h3>
  <p>\[
\begin{aligned}
\tilde{\sigma}_i^2 &= \frac{d_0 \hat{\sigma}_0^2 + d_i
\hat{\sigma}_i^2}{d_0 + d_i}  \\
 &= \left( \frac{d_0}{d_0 + d_i} \right) \hat{\sigma}_0^2 +
\left( \frac{d_i}{d_0 + d_i} \right) \hat{\sigma}_i^2 \\
 &= B \hat{\sigma}_0^2 + (1-B) \hat{\sigma}_i^2
\end{aligned}
\]</p>

  
</section>
<section class='' data-state='' id='slide-30'>
  <h3>Proper hierarchical model</h3>
  <p><img src="img/plate2.png" alt=""></p>

  
</section>
<section class='' data-state='' id='slide-31'>
  <h3>Estimation of hyperparameters</h3>
  <ul>
<li>Need to estimate \(d_0, \hat{\sigma}_0^2\), which control strength and
location of <em>shrinkage</em> or <em>moderation</em></li>
<li>\(d_0, \hat{\sigma}_0^2\) estimated via first two moments of \(\log \hat{\sigma}_i^2\)</li>
<li>(Also need to estimate \(\upsilon_{0}\), another parameter giving
variance of coefficients)</li>
</ul>

  
</section>
<section class='' data-state='' id='slide-32'>
  <h3>limma vs. naive estimators by rank</h3>
  <p><img src="assets/fig/roc3-1.png" alt="plot of chunk roc3"></p>

  
</section>
<section class='' data-state='' id='slide-33'>
  <h3>Operating characteristics (FP &amp; FN)</h3>
  <p><img src="assets/fig/roc4-1.png" alt="plot of chunk roc4"></p>

  
</section>
<section class='' data-state='' id='slide-34'>
  <h3>Summary</h3>
  <ul>
<li>limma provides a hierarchical model for moderation of
variance estimates in the context of linear models</li>
<li>Avoids false positives from under-estimation of variance</li>
<li>Also addresses the gain in degrees of freedom from moderation</li>
</ul>

  
</section>
<section class='' data-state='' id='slide-35'>
  <h3>RNA-seq: counting molecules</h3>
  <p><img src="img/illumina.png" alt=""></p>

  
</section>
<section class='' data-state='' id='slide-36'>
  <h3>RNA-seq: counting molecules</h3>
  <p><br></p>

<pre><code>@SRR1265495.1 1/1
CTTTGCCCGCGTGTCAGACTCCATCCCTCCTCTGCCGCCACCGCAGCAGCCACAGGCAGAGGAGGACGAGGACGACTGGGAATCGTAGGGGGCTCCATGAC
+
CCCFFFFFHGFHHHIJJIJIJJIJJJJJJJGIIJJJIJJJJJJJJEFHDEFFFEECDDDBDD?BB?B@BDDD@;9&lt;BBBDBCB@A2&lt;?BDDDBDBD@CDDC
@SRR1265495.2 2/1
CCTGGCTGTGTCCATGTCAGAGCAATGGCCCAAGTCTGGGTCTGGGGGGGAAGGTGTCATGGAGCCCCCTACGATTCCCAGTCGTCCTCGTCCTCCTCTGC
+
@C@FFFDFHHHHGGIIAGHI9GIIIIIGIIIIIGI@@FHGDDDH@GGIBB05?B&gt;ACCCCCCCB&lt;?BBCCBBBBCBCDC3&gt;AC&lt;BB&lt;?CBBB?&lt;@CCCA:@
@SRR1265495.3 3/1
CTGTGTCCATGTCAGAGCAATGGCCCAAGTCTGGGTCTGGGGGGGAAGGTGTCATGGAGCCCCCTACGATTCCCAGTCGTCCTCGTCCTCCTCTGCCTGTG
+
@C@DFFFFGGHDHIIEGDHCGGHIJGEHIIJIIJIEEHGIGEGDDB@@@CDDDDEDDDDDBDDDDDDDDDBCCDDCCDBBDDDDDB@?@DDDDCDDCC@4&gt;
</code></pre>

<p>for ~30 million reads (often pairs of reads)</p>

  
</section>
<section class='' data-state='' id='slide-37'>
  <h3>RNA-seq: counting molecules</h3>
  <p>Align to genome or transcriptome</p>

<p><img src="img/blat.png" alt=""></p>

  
</section>
<section class='' data-state='' id='slide-38'>
  <h3>RNA-seq: counting molecules</h3>
  <ul>
<li>Now for each gene and each sample we can obtain a <em>count</em>
or <em>estimated count</em> of fragments</li>
<li>Why estimated? Because some fragments cannot be uniquely associated
with genes or isoforms</li>
<li>Fast algorithms for probabilistically assignment:

<ul>
<li><a href="http://www.nature.com/nbt/journal/v34/n5/full/nbt.3519.html">kallisto</a> (2016)</li>
<li><a href="http://www.nature.com/nmeth/journal/v14/n4/abs/nmeth.4197.html">Salmon</a> (2017)</li>
</ul></li>
<li>Assume we have integer counts \(K_{ij}\) of unique fragments or from 
rounding estimated counts</li>
</ul>

  
</section>
<section class='' data-state='' id='slide-39'>
  <h3>Counts</h3>
  <ol>
<li>Either model data with count distributions and inference with GLM

<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pubmed/25516281">DESeq2</a> (2014)</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pubmed/19910308">edgeR</a> (2010)</li>
<li>many more</li>
</ul></li>
<li>Learn weights associated with log normalized counts and use limma

<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pubmed/24485249">limma-voom</a> (2014)</li>
</ul></li>
</ol>

  
</section>
<section class='' data-state='' id='slide-40'>
  <h3>Important for statistical analysis</h3>
  <ul>
<li>Total number of fragments is technical artifact</li>
<li>Heteroskedasticity of counts</li>
<li>Each gene has different variability</li>
</ul>

  
</section>
<section class='' data-state='' id='slide-41'>
  <h3>Total number of fragments</h3>
  <p><img src="assets/fig/totalnumber-1.png" alt="plot of chunk totalnumber"></p>

  
</section>
<section class='' data-state='' id='slide-42'>
  <h3>Sampling fragments</h3>
  <p><img src="assets/fig/sampfrags-1.png" alt="plot of chunk sampfrags"></p>

  
</section>
<section class='' data-state='' id='slide-43'>
  <h3>Poisson across technical replicates</h3>
  <ul>
<li>From <a href="https://www.ncbi.nlm.nih.gov/pubmed?term=20167110">Bullard 2010</a>
take 7 technical replicates</li>
<li>Calculate expected value \(\hat{\lambda}_{ij}\) using DESeq2 norm</li>
<li>\(P(K_{ij} < \hat{\lambda}_{ij})\) assuming \(K_{ij} \sim \textrm{Pois}(\hat{\lambda}_{ij})\)</li>
</ul>

<p><img src="assets/fig/poisson-1.png" alt="plot of chunk poisson"></p>

  
</section>
<section class='' data-state='' id='slide-44'>
  <h3>However, expression not equal across biological replicates</h3>
  <p><img src="img/people.jpg" alt=""></p>

  
</section>
<section class='' data-state='' id='slide-45'>
  <h3>Negative Binomial / Gamma Poisson</h3>
  <p>\[
\begin{aligned}
K_{ij} &\sim \textrm{NB}(\mu_{ij}, \alpha_i) \\
\textrm{Var}(K_{ij}) &= \mu_{ij} + \alpha_i \mu_{ij}^2
\end{aligned}
\]</p>

<p><img src="assets/fig/negbin-1.png" alt="plot of chunk negbin"></p>

  
</section>
<section class='' data-state='' id='slide-46'>
  <h3>NB model for RNA-seq</h3>
  <ul>
<li>Similar to our microarray model for \(X_{ij}\)</li>
<li>Added an \(s_j\) to deal with sequencing depth</li>
</ul>

<p><br></p>

<p>\[
\begin{aligned}
K_{ij} &\sim \textrm{NB}(\mu_{ij}, \alpha_i) \\
\mu_{ij} &= s_j q_{ij} \\ 
\log_2(q_{ij}) &= \beta_{i0}, \quad j \in A \\
\log_2(q_{ij}) &= \beta_{i0} + \delta_i, \quad j \in B
\end{aligned}
\]</p>

<p><br></p>

<p>\(\delta_i \ne 0\) implies DE (differential expression)</p>

  
</section>
<section class='' data-state='' id='slide-47'>
  <h3>Moderation of dispersion</h3>
  <ul>
<li>In <a href="https://www.ncbi.nlm.nih.gov/pubmed/25516281">DESeq2</a>, a prior on \(\log(\alpha_i)\)</li>
<li>Calculate the mean of normalized counts \(\bar{\mu}_i\)</li>
<li>A trend line of dispersion over mean \(\alpha_{tr}(\mu)\)</li>
<li>Width of the prior \(\sigma^2\) estimated via assumption of normal
sampling variance of \(\log(\hat{\alpha}_i)\)</li>
</ul>

<p><br></p>

<p>\[
\log(\alpha_i) \sim N(\log(\alpha_{tr}(\bar{\mu}_i)), \sigma^2)
\]</p>

  
</section>
<section class='' data-state='' id='slide-48'>
  <h3>Moderation of dispersion</h3>
  <p><img src="img/plate3.png" alt=""></p>

  
</section>
<section class='' data-state='' id='slide-49'>
  <h3>Moderation of dispersion</h3>
  <p><img src="assets/fig/disp-1.png" alt="plot of chunk disp"></p>

  
</section>
<section class='' data-state='' id='slide-50'>
  <h3>Evaluate via simulation</h3>
  <p><img src="assets/fig/rnasim-1.png" alt="plot of chunk rnasim"></p>

  
</section>
<section class='' data-state='' id='slide-51'>
  <h3>Summary</h3>
  <ul>
<li>Model for counts similar to the hierarchical linear model,
but constructed for the dispersion parameter</li>
<li>Final dispersion estimate plug-in value in DESeq2, edgeR</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pubmed/23104842">edgeR quasi-likelihood</a> takes into account dispersion estimation uncertainty</li>
<li><a href="https://www.ncbi.nlm.nih.gov/pubmed/24485249">limma-voom</a> uses weights on log normalized counts</li>
<li><a href="https://doi.org/10.1093/bioinformatics/bty895">apeglm</a> uses Cauchy prior (heavy tails) on effect sizes</li>
</ul>

  
</section>
    </div>
  </div>
</body>
  <script src="libraries/frameworks/revealjs/lib/js/head.min.js"></script>
  <script src="libraries/frameworks/revealjs/js/reveal.min.js"></script>
  <script>
  // Full list of configuration options available here:
  // https://github.com/hakimel/reveal.js#configuration
  Reveal.initialize({
    controls: true,
    progress: true,
    history: true,
    center: true,
    theme: Reveal.getQueryHash().theme || 'serif', 
    transition: Reveal.getQueryHash().transition || 'none', 
    dependencies: [
    // Cross-browser shim that fully implements classList -
    // https://github.com/eligrey/classList.js/
      { src: 'libraries/frameworks/revealjs/lib/js/classList.js', condition: function() { return !document.body.classList;}},
      // Zoom in and out with Alt+click
      { src: 'libraries/frameworks/revealjs/plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
      // Speaker notes
      { src: 'libraries/frameworks/revealjs/plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } },
      // Remote control your reveal.js presentation using a touch device
      //{ src: 'libraries/frameworks/revealjs/plugin/remotes/remotes.js', async: true, condition: function() { return !!document.body.classList; } }
      ]
  });
  </script>  <!-- MathJax: Fall back to local if CDN offline but local image fonts are not supported (saves >100MB) -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
  });
</script>
<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<!-- <script src="https://c328740.ssl.cf1.rackcdn.com/mathjax/2.0-latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script> -->

<script>window.MathJax || document.write('<script type="text/x-mathjax-config">MathJax.Hub.Config({"HTML-CSS":{imageFont:null}});<\/script><script src="libraries/widgets/mathjax/MathJax.js?config=TeX-AMS-MML_HTMLorMML"><\/script>')
</script>

<!-- LOAD HIGHLIGHTER JS FILES -->
<script src="libraries/highlighters/highlight.js/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<!-- DONE LOADING HIGHLIGHTER JS FILES -->
 

</html>